handlers.go

```diff
@@ -593,13 +603,52 @@ func (s *Server) SaveBlob(w http.ResponseWriter, r *http.Request) {
 		return
 	}
 
+	if err := os.Rename(tf.Name(), path); err != nil {
+		_ = os.Remove(tf.Name())
+		if s.MaxRepoSize > 0 {
+			s.incrementRepoSpaceUsage(-written)
+		}
+		if s.Debug {
+			log.Print(err)
+		}
+		http.Error(w, http.StatusText(http.StatusInternalServerError), http.StatusInternalServerError)
+		return
+	}
+
+	if err := syncDir(filepath.Dir(path)); err != nil {
+		// Don't call os.Remove(path) as this is prone to race conditions with parallel upload retries
+		if s.Debug {
+			log.Print(err)
+		}
+		http.Error(w, http.StatusText(http.StatusInternalServerError), http.StatusInternalServerError)
+		return
+	}
+
 	if s.Prometheus {
 		labels := s.getMetricLabels(r)
 		metricBlobWriteTotal.With(labels).Inc()
 		metricBlobWriteBytesTotal.With(labels).Add(float64(written))
 	}
 }
 
+func syncDir(dirname string) error {
+	if runtime.GOOS == "windows" {
+		// syncing a directory is not possible on windows
+		return nil
+	}
+
+	dir, err := os.Open(dirname)
+	if err != nil {
+		return err
+	}
+	err = dir.Sync()
```

Comment:
Calling this for every write looks expensive. Why add this? A directory entry will never be partly written on a well-designed modern filesystem.

Reply:
Without the sync call a valid filesystem behavior would be to loose the directory entry altogether if the system crashes after the rest-server has accepted the file. So the intention here is to make sure that not only the file content, but also the directory entry survives once the rest-server confirms that the upload was successful.

Reply:
OK, I don't mind. I'm just not sure that this will actually be effective, but in this case it should not hurt to at least try the best we can.

I was thinking about this LWN article, but that is mostly about how I/O errors during fsync are handled: https://lwn.net/Articles/752063/

Reply:
More about performance impact: http://oldblog.antirez.com/post/fsync-different-thread-useless.html

Reply:
I guess if syncing a directory to disk fails, then we could also let the rest-server panic and exit as there's no good way we can fix that mess. Strictly speaking it would be best to just reboot the whole system in that case, but that's rather out of scope for the rest-server.

Although the current way of telling the client that the disk write failed should solve most problems. The client would have to retry the request in this case. And if that manages to successfully flush the directory for the retried file upload, then there's a good chance that the directory is again stored properly on disk.

> More about performance impact: http://oldblog.antirez.com/post/fsync-different-thread-useless.html

That blog post is unfortunately rather old and probably still used ext3 which was especially slow when flushing. But that's speculation as the blog post neither mentions the used kernel, filesystem, HDD, etc. And I doubt that the magically faster `O_SYNC` numbers in that blog post are correct. After all these should be roughly equivalent to calling write()+fsync().